### PR TITLE
Port THS to ATen.

### PR DESCRIPTION
**Pull request by ezyang**
The basic structure of the patch:

- All kernels in aten/src/THS got rewritten as native
  functions in aten/src/ATen/native/sparse

  I took the liberty to rename some of the kernels,
  opting for a longer, more transparent names than
  things like 'spaddcmul'.

- Instead of holding fields for sparse tensor in the TH
  C struct THSTensor, they are now held in a C++ class
  SparseTensorImpl (this explains why I had to do this
  all in one go; I can't have *two* reps for sparse
  tensors!)

  Along the way, we change a key internal representation
  invariant: an "empty" sparse tensor has dimI == 1 and
  dimV == 0 (this is different from dimI == 0 and dimV == 0
  we had before); this ensures that we maintain the invariant
  that dim == dimI + dimV.  "Scalar" sparse tensors are
  made illegal, because there really is no way to properly
  express them in COO format.

- Because we haven't ported THCS or any of the traditional
  dense TH implementations, there is a new set of adapter
  functions in native/LegacyBridge.cpp exclusively devoted
  to deciding whether or not to go to the new native implementation
  or back to the legacy TH binding (prefixed with th_).
  The intent is that when everything gets ported, we can
  delete this file.

- I've kept the stubs for all the THS functions, but they now all
  error if you try to actually call them.  Eventually, we should
  replace these with calls to ATen so that everything keeps
  working.

- I gobbled up SparseMM (SparseMM.cpp is no more). It was tasty.

There are some miscellaneous improvements which were needed for other
changes in this patch:

- There is now AT_FORALL_SCALAR_TYPES_EXCEPT_HALF, which does what
  it says on the tin.

- axpy templated function moved to BlasUtils.h, there's a new macro
  which lets you easily forward to all of the TH functions. We also expose
  THBlas_(copy).  I'm not terribly pleased with these functions but
  they seem to serve a purpose they need.

- New method on Tensor to get TensorImpl*, unsafeGetTensorImpl

- accessor() is now this-const, since const-correctness on Tensor is a lie

- New toSparse()/toDense() methods on Type; now you can call these
  directly without having to manually apply at::toSparse/toDense
  on the Backend and then running toBackend yourself.

Changes to the kernels:

- Previously, the whole body of all kernels was compiled for
  every supported scalar type.  In our new implementation,
  the scalar dispatch has been pushed into the smallest extent
  which (1) is not in a type loop and (2) requires statically
  knowing the scalar type.  These sites all use
  AT_DISPATCH_ALL_TYPES.  I tried to use lambdas as much as
  possible, but sometimes it was not possible when a OpenMP
  pragma was used.

- Anywhere we tested if the nDimension of a tensor was zero,
  we replaced with a test that numel is zero.  Because, as we
  known, nDimension of zero-size tensors in TH is zero, and
  that's wrong wrong wrong (and not done this way in ATen).

Some subtleties:

- Places where previously fastget1d was used, I now use a
  TensorAccessor.  However, you have to be careful about grabbing
  the accessor, because sometimes you will be accessor'ing
  indices/values and they are empty, which means they will
  be *1D* ("oh, aren't indices always 2D?" Nope. Nyet.)
  So, essentially, it is only safe to grab an accessor *after*
  you have checked that nnz != 0.  All of these shenanigans
  will go away when we properly support zero-size dimensions.

  A few places, we test for this case just by wrapping the loop
  in a conditional on nnz.  Some other places this is not so easy,
  so we instead short-circuit the function with a special case for
  when nnz == 0 (usually, these implementations are degenerate).

- There is a very subtle but important difference between
  _sparse_get_impl(self)->indices() and self._indices();
  the latter may return a view!  This is because nnz is
  not guaranteed to match the dimensions of indices/values;
  you can "truncate" a sparse tensor by setting the nnz.
  Actually, I think this is not a good idea and we should
  enforce a stronger invariant, but for this patch I slavishly
  adhere to the old ways, and as such I have to be very
  careful if I want to resize something, I had better use
  the former and not the latter.

- I had to reimplement broadcasting by hand (thus the s_
  and non-s_ functions in the sparse native files).  There
  is a very important distinction between foo_out and foo_,
  so it is important that the LegacyBridge function always
  call to the lower layer, and not try to avoid boilerplate
  by calling to another LegacyBridge function first.
  I did NOT put broadcasting in LegacyBridge (even though,
  ultimately, that's where it must live), because the th_
  functions which are invoked from LegacyBridge handle
  broadcasting themselves, and I don't want to broadcast
  twice.

- Sparse function MUST explicitly specify the Type they
  dispatch from, otherwise Variable wrapping/unwrapping will
  not work correctly.  If you use _get_sparse_impl, that is
  sufficient to levy this requirement.

- The "has native" tests in LegacyBridge.cpp are not 100%,
  because some of the functions are mixed dense-sparse functions,
  and so you can't just say, "Oh, if it's sparse and CPU, call
  the native sparse implementation."  This is handled on a
  case by case basis.  There is some especially complex
  logic for add(), which has dense-dense, sparse-sparse
  and dense-sparse implementations.

- I added some uses of SparseTensorRef in native_functions.yaml,
  but you will notice that these are all on native_* functions,
  and not the actual, top-level functions.  So the SparseTensorRef
  is purely documentary (helping you not call the wrong overload)
  but there is no magic; we do the wrapping ourselves the hard
  way. (This is in constrast to the TH binding code which is magical.)
  Except for _sparse_mask; _sparse_mask is magical.

- There is a raw_copy_sparse_ method, which is really my way of
  getting around the fact that copy_ has never been implemented
  for sparse tensors (even before this patch), but there IS a
  super secret, internal way of doing these copies that the THS
  code used, and which I needed to get my hands on when I did this
  port.  We should refactor so that either (a) copy_ does support
  sparse-sparse copy natively, or (b) we do this other ways.

- Irritatingly, I must explicitly resize_as_ before copy_ into
  a tensor.  This was not the case with THTensor_(copy) but I don't
  have any direct binding that doesn't have this requirement.

- For some reason, the sparse tensor constructor accepts a scalar
  tensor for the values tensor.  This is kind of weird because
  you always need an nnz-dimension.  However, the old code supported
  this and just expanded it into a 1D size 0 tensor; so we need some
  explicit code to do this.

There are maybe a bit more AT_ASSERTs in some of the kernels
than is wise.  I added them all when I was debugging and was
loathe to remove them.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

TODO: I claim that I have "error"ed out all of the old THS functions, but this has not actually been done yet except for one function.

### CODE REVIEW

**Comment by ezyang** [comment]
I know why the broadcast tests are failing (the broadcast calls are behind a dispatch: SparseCPU, so they aren't interpreted as part pieces I need to autograd through) but there is some special casing for this situation in TH which I can't seem to quite replicate when some overloads are native while others are TH. I'll try to work it out tomorrow.

**Comment by ezyang** [comment]
There are some very uncomfortable hacks in the last set of commits. Please review with care.

File: aten/src/ATen/native/native_functions.yaml
Comment (ezyang) [reply aten/src/ATen/native/native_functions.yaml]: That's not unreasonable, although it wouldn't be enough for this case, because neither argument is the actual type you want the dispatch to happen on in the end.

Actually, when C10 dispatcher lands none of this will matter because you won't have to pick an argument to dispatch on.

File: aten/src/ATen/SparseTensorImpl.cpp
Comment (gchanan) [inline aten/src/ATen/SparseTensorImpl.cpp]: will values.size(0) not be 0 in the empty case?

File: aten/src/ATen/SparseTensorImpl.h
Comment (gchanan) [inline aten/src/ATen/SparseTensorImpl.h]: should we rename these?

File: aten/src/ATen/SparseTensorImpl.h
Comment (gchanan) [inline aten/src/ATen/SparseTensorImpl.h]: what does ordered mean here?  do you just mean coalesced?

File: aten/src/ATen/SparseTensorImpl.h
Comment (gchanan) [inline aten/src/ATen/SparseTensorImpl.h]: you could future proof this code with USE_TH_ZERO_SIZE_DIM as in https://github.com/pytorch/pytorch/pull/8468 if you want.

File: aten/src/ATen/ScalarType.h
Comment (ezyang) [reply aten/src/ATen/ScalarType.h]: Sure I'll move it to BlasUtils.h

File: aten/src/ATen/SparseTensorImpl.cpp
Comment (ezyang) [reply aten/src/ATen/SparseTensorImpl.cpp]:
```
// SparseTensorImpl defaults to a [0] size tensor with one sparse dimension and
// no dense dimensions.  This is kind of arbitrary but we want an empty tensor
// to have one dimension (zero dimensions means its scalar), and we have
// the invariant that dim == dimI + dimV, so *one* of the inner dimensions
// has to be one.
```

File: aten/src/ATen/SparseTensorImpl.cpp
Comment (ezyang) [reply aten/src/ATen/SparseTensorImpl.cpp]: Yes, but it's an accident. I think it's clearer this way. I added this comment:

```
  // TODO: Eliminate this ternary when we handle size zero dimensions.
  // (Actually, this will "accidentally" work today because all zero-size
  // tensors have size [0], and so you'll get 0 when empty is zero; but it's
  // more explicit this way.)
```

File: aten/src/ATen/SparseTensorImpl.h
Comment (ezyang) [reply aten/src/ATen/SparseTensorImpl.h]: Maybe. Note that this name is exposed semi-publically as `_dimI()` method. Do you have good suggestions? `_sparseDims` and `_denseDims`?

File: aten/src/ATen/SparseTensorImpl.h
Comment (ezyang) [reply aten/src/ATen/SparseTensorImpl.h]: yep, coalesced

File: aten/src/ATen/SparseTensorImpl.h
Comment (ezyang) [reply aten/src/ATen/SparseTensorImpl.h]: sure, if you get the tests to pass before I merge this XD

File: aten/src/ATen/SparseTensorImpl.h
Comment (gchanan) [reply aten/src/ATen/SparseTensorImpl.h]: we can still keep them around for legacy reasons.  I like the names you suggested.

File: aten/src/ATen/SparseTensorImpl.cpp
Comment (gchanan) [inline aten/src/ATen/SparseTensorImpl.cpp]: I find this comment a little confusing; can you structure it as:
- This is what it _should_ be (i.e.size [1,0])
- Since we don't have support for that, this is what we chose to do instead

File: aten/src/ATen/SparseTensorImpl.cpp
Comment (gchanan) [inline aten/src/ATen/SparseTensorImpl.cpp]: I don't think this is right; the requirement in dense is that numel() == 1, i.e. this is basically a conversion.  We should do the same thing here.

File: aten/src/ATen/SparseTensorImpl.h
Comment (gchanan) [inline aten/src/ATen/SparseTensorImpl.h]: maybe, as in https://github.com/pytorch/pytorch/pull/8468, we should add `legacy` to the name?

File: aten/src/ATen/copy_wrapper.py
Comment (gchanan) [inline aten/src/ATen/copy_wrapper.py]: I'm confused -- how is this function useful now?

File: aten/src/ATen/native/BlasUtils.h
Comment (gchanan) [inline aten/src/ATen/native/BlasUtils.h]: Could just move this specialization into non-generic TH (hpp). I did something similar with the Copy functions in THC, that way you don't have to invent new ATen APIs.  But this looks fine to me if you want to keep it.

File: aten/src/ATen/native/LegacyBridge.cpp
Comment (gchanan) [inline aten/src/ATen/native/LegacyBridge.cpp]: nit: _type_has_native(self.type())

File: aten/src/ATen/native/LegacyBridge.cpp
Comment (gchanan) [inline aten/src/ATen/native/LegacyBridge.cpp]: could you explain how it's asymmetric?

File: aten/src/ATen/native/LegacyBridge.cpp
Comment (gchanan) [inline aten/src/ATen/native/LegacyBridge.cpp]: this actually works for sparse :)?  I'm pleasantly surprised.

File: aten/src/ATen/native/LegacyBridge.cpp
Comment (gchanan) [inline aten/src/ATen/native/LegacyBridge.cpp]: I'll admit I don't understand why this calls `native_add_out` but the above calls `s_native_add_out` -- the SparseTensorRef stuff makes sense though.

File: aten/src/ATen/native/README.md
Comment (gchanan) [inline aten/src/ATen/native/README.md]: can you not used the word `real` -- that has another meaning an a similar context :).  Maybe something like "an actual", "a realized" or similar.

File: aten/src/ATen/native/README.md
Comment (gchanan) [inline aten/src/ATen/native/README.md]: nice!

File: aten/src/ATen/native/native_functions.yaml
Comment (gchanan) [reply aten/src/ATen/native/native_functions.yaml]: Also I think just taking a Type or ScalarType would work (as you suggest) -- you can do all the same logic that is implemented in tensor_new here.

File: aten/src/ATen/native/native_functions.yaml
Comment (gchanan) [inline aten/src/ATen/native/native_functions.yaml]: wait -- what does this mean?

File: aten/src/ATen/native/sparse/SparseTensor.cpp
Comment (gchanan) [inline aten/src/ATen/native/sparse/SparseTensor.cpp]: I think I called this `is_congruent` somewhere.

File: aten/src/THS/generic/THSTensor.cpp
Comment (gchanan) [inline aten/src/THS/generic/THSTensor.cpp]: why did we keep these?  I agree we could eventually have these call ATen -- but what's the point of not just failing at compile time?

File: test/test_jit.py
Comment (gchanan) [inline test/test_jit.py]: do you want to print the trace in the test?  It's fine, just curious.

File: tools/autograd/gen_python_functions.py
Comment (gchanan) [inline tools/autograd/gen_python_functions.py]: why do you actually need this?  I don't really know what `dynamic_type` is trying to get across here, so it's not obvious to me that it's wrong.  Can you just change is_coord_smaller to do what you want?

File: tools/autograd/gen_variable_type.py
Comment (gchanan) [inline tools/autograd/gen_variable_type.py]: is there a plan to get rid of these?

File: tools/autograd/gen_variable_type.py
Comment (gchanan) [inline tools/autograd/gen_variable_type.py]: what's the plan moving forward with this?  How difficult is it to do something like 'we know this is a native function calling into a non-native function that does broadcasting...'.

File: aten/src/ATen/SparseTensorImpl.cpp
Comment (ezyang) [reply aten/src/ATen/SparseTensorImpl.cpp]:
```
// An empty dense tensor defaults to a 1-dimensional tensor of size [0]
// (recall, it is not a 0-dimensional tensor, because such a tensor would
// a scalar and have one element)
//
// Thus, an empty sparse tensor should be a 1-dimensional tensor of size [0].
// Furthermore, we have dim == sparseDims + denseDims; since this is a sparse
// tensor, let us say that an empty sparse tensor has sparseDims == 1 and
// denseDims == 0.  (There is a degree of freedom here, but given that this
// is a sparse dimension, it seems reasonable to demand that sparseDims > 0).
//
// In an ideal world, this would then mean we allocate a [1,0] size indices
// tensor and a [0] size values tensor for such an empty tensor.  However,
// we don't currently support zero-size dimensions, so we can't actually
// do this; so we just allocate zero-size tensors for everything.
```

Actually, there is a bit in the comment defending why we should default sparseDims == 1, denseDims == 0, I reworded it for clarity.

File: aten/src/ATen/SparseTensorImpl.cpp
Comment (ezyang) [reply aten/src/ATen/SparseTensorImpl.cpp]: It appears to be broken in master (I see localScalar is used in `is_nonzero`):

```
>>> import torch
>>> i = torch.LongTensor([[0, 2], [1, 0], [1, 2]])
>>> v = torch.FloatTensor([3,      4,      5    ])
>>> y = torch.sparse.FloatTensor(i.t(), v)
>>> y.is_nonzero()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: numel is not implemented for type torch.sparse.FloatTensor
```

but seems reasonable enough to fix.

Comment (ezyang) [reply aten/src/ATen/SparseTensorImpl.cpp]: Fixed a lot of it; while working on it I realized that "scalar" sparse tensors actually make sense as a coherent concept, but I didn't go far enough to fix code to handle this correctly.

File: aten/src/ATen/SparseTensorImpl.h
Comment (ezyang) [reply aten/src/ATen/SparseTensorImpl.h]: agreed

Comment (ezyang) [reply aten/src/ATen/SparseTensorImpl.h]: Oh fuck it, I'm going to make scalar sparse tensors work

File: aten/src/ATen/copy_wrapper.py
Comment (ezyang) [reply aten/src/ATen/copy_wrapper.py]: There's some hidden code that you can't see here which overrides `_s_copy_from` for CUDA tensors. So the general strategy is `_s_copy_from` is called from `s_copy` if a cpu-cuda conversion is involved; it's the double dispatch function.

File: aten/src/ATen/native/BlasUtils.h
Comment (ezyang) [reply aten/src/ATen/native/BlasUtils.h]: Moving into THC sounds fine.

File: aten/src/ATen/native/LegacyBridge.cpp
Comment (ezyang) [reply aten/src/ATen/native/LegacyBridge.cpp]:
```
// Note [Multiple dispatch to sparse]
// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
// In an ideal world, we would use direct support for multiple dispatch to
// say that add(Dense, Dense) should dispatch to one function, while
// add(Dense, Sparse) should dispatch to another function.
//
// In a world where we only have single dispatch, we can single dispatch on
// the first function, and then do an is_sparse() test on the second argument
// to direct ourselves to the correct argument.
//
// We are in neither of those worlds.  Instead, we have a th_add function
// which has legacy implementations in the single dispatch world, BUT our
// actual add function needs to call s_native_add if the function *would have*
// utilized a kernel implemented in THS (CPU sparse tensors).
//
// th_add is "good old single dispatch" which internally handles the is_sparse()
// test and also handles broadcasting.  s_native_add works asymmetrically:
// it doesn't handle broadcasting at all, and it ASSUMES that the relevant
// argument is a CPU sparse tensor.  Why the asymmetry?  It turns out it is not
// so easy to figure out if a kernel is implemented in THS; it's not as simple
// as testing if the first argument is CPU and sparse, because, e.g.,
// in add(Dense, Sparse), the sparse kernel is in the second argument.  So,
// the trampoline function is going to know about the overloads *anyway*; it
// might as well also handle is_sparse() and broadcasting while it's at it.
//
// Why not change TH to follow this new scheme?  We could... but since it's
// all going away when we finish porting the TH functions to ATen, we haven't
```

New comment

File: aten/src/ATen/native/LegacyBridge.cpp
Comment (ezyang) [reply aten/src/ATen/native/LegacyBridge.cpp]: I bet it only actually works if it's a dense tensor that gets broadcasted, cuz I'm pretty sure we don't have expand on sparse :) (in any case the tests didn't see anything wrong).

**Review by gchanan — APPROVED** [review]
I don't think I fully understand what this will look like once everything is ported over, but this looks good to merge if you want.

File: aten/src/ATen/native/LegacyBridge.cpp
Comment (ezyang) [reply aten/src/ATen/native/LegacyBridge.cpp]: It's pretty simple, actually; in the old Declarations.cwrap, there was no broadcasting enabled for dense-sparse addition, so I mindlessly copied the behavior over.

File: aten/src/ATen/native/native_functions.yaml
Comment (ezyang) [reply aten/src/ATen/native/native_functions.yaml]: `variants: []` means, don't generate `at::native_sparse_coo_tensor` or `Tensor:: native_sparse_coo_tensor`, but do create a method on `Type`. But I think I'm going to see if I can just make these take a `Type`.

File: aten/src/THS/generic/THSTensor.cpp
Comment (ezyang) [reply aten/src/THS/generic/THSTensor.cpp]: I kept it so that, if we eventually have these call ATen, all of the templates will still be here so it's an easier patch to do. Otherwise whoever does it has to dig out the old files, wire them all back up, and then start doing replacements.

File: test/test_jit.py
Comment (ezyang) [reply test/test_jit.py]: Nope, it's an accident

File: tools/autograd/gen_variable_type.py
Comment (ezyang) [reply tools/autograd/gen_variable_type.py]: Yeah, the plan on record is https://github.com/pytorch/pytorch/issues/8495

File: tools/autograd/gen_python_functions.py
Comment (ezyang) [reply tools/autograd/gen_python_functions.py]: I can't only change `is_coord_smaller`, because there are equality tests done on `dynamic_type`, and if `real` and `Scalar` don't compare as equal I get the wrong sort. Maybe I should just fix this at the source and make sure everything works out OK. What do you think?

File: aten/src/ATen/native/native_functions.yaml
Comment (ezyang) [reply aten/src/ATen/native/native_functions.yaml]: I attempted this patch, but I triggered a codegen bug which I don't feel like fixing:

```
In file included from torch/csrc/autograd/generated/python_torch_functions.cpp:23:0:
torch/csrc/autograd/generated/python_torch_functions_dispatch.h: In function ‘at::Tensor torch::autograd::dispatch__native_sparse_co
o_tensor_unsafe(const at::Tensor&, const at::Tensor&, at::IntList)’:
torch/csrc/autograd/generated/python_torch_functions_dispatch.h:584:25: error: ‘dtype’ was not declared in this scope
   maybe_initialize_cuda(dtype);
```

File: tools/autograd/gen_python_functions.py
Comment (gchanan) [reply tools/autograd/gen_python_functions.py]: not sure what you mean by at the source, do you mean so the native and non-native declarations entries are the same?  I find it hard to argue about this without really understanding what `dynamic_type` is supposed to represent.

File: aten/src/ATen/native/native_functions.yaml
Comment (ezyang) [reply aten/src/ATen/native/native_functions.yaml]: https://github.com/ezyang/pytorch-1/tree/snapshot/taking-type-for-sparse-coo-tensor

File: tools/autograd/gen_variable_type.py
Comment (ezyang) [reply tools/autograd/gen_variable_type.py]: Also #8495

File: aten/src/ATen/native/native_functions.yaml
Comment (gchanan) [reply aten/src/ATen/native/native_functions.yaml]: I think we should consider methods on types to be an implementation detail, it doesn't match up with the (non-legacy) pytorch API.

File: tools/autograd/gen_python_functions.py
Comment (ezyang) [reply tools/autograd/gen_python_functions.py]: Yep, that's what I mean. Pinging @zdevito about the meaning of `dynamic_type` versus `type`.

**Pull request merged**
